include/seastar/core/semaphore.hh

```diff
@@ -276,17 +391,20 @@ public:
         if (_state->ex) {
             return make_exception_future(_state->ex);
         }
-        try {
-            entry& e = _wait_list.emplace_back(promise<>(), nr);
-            auto f = e.pr.get_future();
-            if (timeout != time_point::max()) {
-                e.timer.emplace(timeout);
-                abort_source& as = e.timer->abort_source();
-                _wait_list.make_back_abortable(as);
+        if (Clock::now() >= timeout) {
+            try {
+                return make_exception_future(this->timeout());
+            } catch (...) {
+                return make_exception_future(semaphore_timed_out());
```

Comment:
`timeout()` doesn't throw in seastar but the user may use their own ExceptionFactory and implement a throwing timeout() method (e.g. it could format/allocate a string).
Once we move the baseline support to C++20 we can add concepts requiring it to be noexcept.